### ISSUE TITLE
Updating image source/tag in parameters can break workspace updates

### ISSUE THREAD

**Issue by bpmct** [coder/coder#9163]
## Background

I highly recommend watching [my Loom video](https://www.loom.com/share/91bb44b685f74ea8ac897fba25db73dc?sid=3c709744-cc5f-4028-bebd-3f02ec010a53) that describes the problem.

The problem was initially described here #7442 but I figured I would create a new issue to describe the problem, share a couple of potential ways we can fix this in the product, and share workarounds!

## Must have

- [ ] When parameter values change and there is a validation error, the user should see a pop-up to pick an appropriate value for the new version. Right now, templates get bricked
- [ ] Documentation for the proper way to update image sources and tags via parameters

## Should have

- [ ] Some way to update image source / tag without using `locals` as a dictionary (e.g. we only save the name or add another value to save instead of the `value` in the DB. I'm not really sure the best way to do this, but this is a common use case and the `locals` workaround feels a bit hacky to me.

## The `locals` workaround

If you are a template admin, you can do this to safely update your workspaces without causing workspaces to get un-updatable:

```hcl
variable "go_image" {
}
variable "java_image" {
}

locals {
  images = {
    "go": var.go_image,
    "java": var.java_image
  }
}

data "coder_parameter" "image" {
  name = "Image"
  default = "go"
  type = "string"
  option {
    name = "Golang"
    value = "go"
  }
  option {
    name = "Java"
    value = "java"
  }
  mutable = "true"
}

resource "docker_container" "workspace" {
  count = data.coder_workspace.me.start_count
  image = local.images[data.coder_parameter.image.value]
  # 
}
```

Example commands:

```sh
# Create the template
coder templates push --var go_image="mcr.microsoft.com/devcontainers/go:1.20" --var java_image="mcr.microsoft.com/devcontainers/java:8" -y

# Update all java workspaces to Java 11
coder templates push --var go_image="mcr.microsoft.com/devcontainers/go:1.20" --var java_image="mcr.microsoft.com/devcontainers/java:11" -y
```

**Comment by mtojek** [comment]
>  When parameter values change and there is a validation error, the user should see a pop-up to pick an appropriate value for the new version. Right now, workspaces get stuck and are unable to update

This is a special case, and definitely a bug which is the outcome of the current design. I will focus on this in the first place.

**Comment by mtojek** [comment]
At the moment, on the workspace update, the Coder (CLI/UI) prompts the user to select the new parameter option if the select option from the last build is incompatible. Anyway, the workspace user will not be stuck anymore.

<img width="586" alt="Screenshot 2023-08-24 at 11 20 01" src="https://github.com/coder/coder/assets/14044910/84275967-89b4-469b-87d5-fa205c949b1d">
<img width="618" alt="Screenshot 2023-08-24 at 11 20 16" src="https://github.com/coder/coder/assets/14044910/52d5466d-442d-4565-82e1-23507186e3ac">

>  Some way to update image source / tag without using locals as a dictionary.

> Some ideas: we only save parameter name in the DB. the value is interpreted from the new version or add another value (e.g. slug) to save instead of the value in the DB.

Is the current situation good enough, or should we iterate on this? Honestly, I wouldn't _silently_ look up the previous option as it might be problematic as well. Consider the following case:

Before: 

```
cat=aaa
dog=bbb
cow=ccc
```

After:
```
cat=aaa
dog=eee
horse=bbb-continue
cow=ccc
```

As you can see the value `bbb` disappeared, and if we try to automatically match on option name, `dog`, the user will end up with value `eee`, where the expectation is to go with `bbb-continue`.

**Comment by johnstcn** [comment]
Silently updating a parameter and trying to guess the correct option smells too much like magic to me. Prompting the user is simpler and maintains their agency.

**Comment by bpmct** [comment]
Hmm... I do believe we need to iterate on this, or at the very least document the `locals` approach. 

This "updating image tag" use case is going to be a common one: an admin offers 3 different images to developers and wants to keep image tags in sync with template versions so the user is promoted to update their workspace when a new image tag comes out. When a developer rolls back to an old template version, they will have a template that "just worked" at that point in time by using the older image.

The `locals` approach I shared feels like pretty hacky to recommend and difficult to maintain. It feels like we actually solved this issue in parameters itself by adding a `display_name` which allows the admin to freely iterate on the display name of the parameter in the UI. If the admin changes the `name` of a parameter, they know that users will be re-prompted, but changing the display name is safe. We did a similar thing with `coder_apps` and slugs. For example, the destination of a `coder_app` can silently version to version but it display looks the same in the dashboard. 

Alternatively, I found a cleaner way to prompt the user for an image they want to use. Well, cleaner is an overstatement. It trades duplication for some complex Terraform functions.

```hcl
variable "go_image" {}
variable "java_image" {}

locals {
  images = {
    "Golang" = var.go_image,
    "Java"   = var.java_image
    # Add more as needed
  }
}

data "coder_parameter" "image" {
  name    = "Image"
  type    = "string"
  default = "golang"

  dynamic "option" {
    for_each = [for k in keys(local.images): { name = k, value = lower(k) }]
    content {
      name  = option.value.name
      value = option.value.value
    }
  }

  mutable = "true"
}

resource "docker_container" "workspace" {
  count = data.coder_workspace.me.start_count

  image = local.images[title(data.coder_parameter.image.value)]
  # ...
```

I suppose we could abstract this complexity from the template by providing and maintaining an "image picker" module or something down the road (we plan on making and publishing modules in the template registry). I am mostly trying to optimize for cases where the developer picked a value, 6 months later the admin wants to silently migrate users who picked a specific parameter, then users are re-prompted out of nowhere and may not remember what they picked or what the difference is ("i picked java, why is it asking me again?!"). 

TLDR: I suppose I am OK kicking the can on this one as long as we document your current recommended way of doing this. I think that silently updating the behavior of a parameter without re-prompting the user is a legitimate use case.

**Comment by mtojek** [comment]
> B) document the locals approach.

The `locals` approach does not have to be considered a hack. From the workspace user perspective: _I'd like to use the latest Java image (`java:latest`). I don't care which version is it._
 
Technically, "latest" would be an alias pointing to the TF variable. Of course I can document "locals" 👍

**Comment by spikecurtis** [comment]
Updating the parameter value based on the `Option` "name" or a new "slug" value makes a lot of assumptions about intent.

Counterexample e.g.: 

```hcl
data "coder_parameter" "image" {
  name = "Image"
  type = "string"
  option {
    name = "latest"
    value = "spikecurtis/workspace:aaaa"
  }
  option {
    name = "july"
    value = "spikecurtis/workspace:bbbb"
  }
  mutable = "true"
}
```

When I built my workspace the first time in July, tag `bbbb` _was_ the `latest`, and I didn't have the option to explicitly pick `july`.  Does this result in `bbbb` because that passes validation, or `aaaa` because the transform happens before validation?  Confusing!

Also, trying to update based on names or slugs only works for multiple-choice parameters.

I think the simplest thing to understand is that Coder always stores the parameter _values_ from a build.  If you upgrade template versions, and the _values_ no longer validate, we prompt.  (And fix this wherever it isn't true today.)

Documenting the locals approach and converting our example templates to use it will go a long way to making this sharp edge fairly rare.

**Comment by bpmct** [comment]
Cool. Let's do `locals` then 👍🏼

**Comment by spikecurtis** [comment]
> @bpmct - is the recommendation here to use `locals` for the `default` value? e.g.
> 
> ```
> data "coder_parameter" "base_image" {
>   name        = "base_image"
>   type        = "string"
>   description = <<-EOF
>   Base image to use for workspace
>   EOF
>   default = local.image
>   mutable = true
>   order   = 3
> }
> ```
> 
> a customer is reporting that updating the `default` value is not being reflected in a workspace update.

no, in that example you are setting the default to the contents of `local.image`.

The "locals" approach described in this issue is to set the default to something like the literal `"latest"`, but then do a lookup on that key against a local dictionary to get the actual image.  That way, when select `"latest"` as the parameter, it can _mean_ different images in different template versions, but the contents of the _parameter value_ doesn't change.
